### PR TITLE
use GtkDialog

### PR DESCRIPTION
**Pull request by wdlkmpx**
to make things compatible with gtk3, use gtk_dialog_add_button() and avoid dialog->action_area

### CODE REVIEW

File: src/gtkutils.c [hunk 1]
```diff
@@ -35,50 +35,71 @@ GtkWidget *ut_window_new(gchar *title, gchar *wmname, gchar *wmclass,
...
-   gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(button), active);
-   gtk_box_pack_start(GTK_BOX(parent), button, FALSE, FALSE, 0);
+   if (main_vbox) {
+      *main_vbox = gtk_dialog_get_content_area (GTK_DIALOG (dialog));
+      /* padding */
+      gtk_container_set_border_width (GTK_CONTAINER (*main_vbox), 4);
+   }
 
-   return button;
+   return (dialog);
```
Comment (emorozov) [inline src/gtkutils.c]: Hmm, isn't `return` a statement in C? Why extra parens if it's not a function call.

File: src/gtkutils.c [hunk 2]
```diff
@@ -35,50 +35,71 @@ GtkWidget *ut_window_new(gchar *title, gchar *wmname, gchar *wmclass,
...
-   g_signal_connect(G_OBJECT(button), "clicked", func, func_data);
-   gtk_box_pack_start(GTK_BOX(parent), button, TRUE, TRUE, 0);
-   gtk_widget_set_can_default (button, TRUE);
+   GtkWidget * frame_vbox = gtk_vbox_new (FALSE, 5);
+   gtk_container_add (GTK_CONTAINER (frame), frame_vbox);
+   /* padding */
+   gtk_container_set_border_width (GTK_CONTAINER (frame_vbox), 5);
 
-   return button;
+   return (frame_vbox);
```
Comment (emorozov) [inline src/gtkutils.c]: Yes, I've double-checked and `return` is a statement in C, hence it doesn't require parens around return value. I do not insist, but I like when there're no extra unneeded parens in the code. It makes easier to distinguish statements vs function calls, for example.

File: src/preferences.c [hunk 3]
```diff
@@ -108,22 +98,30 @@ void preferences_apply(GtkWidget *widget, gpointer data)
...
-   gtk_widget_destroy(buttons[PR_DIALOG]);
+static void prefs_dialog_response (GtkDialog * dlg, int response, gpointer user_data)
+{
+   switch (response)
+   {
+      case GTK_RESPONSE_APPLY:
+         preferences_apply (user_data);
+         g_signal_stop_emission_by_name (dlg, "response");
+         return;
+         break;
```
Comment (emorozov) [inline src/preferences.c]: Is `break` here necessary? My C knowledge is very limited but it's a no-op here and as far as I remember C doesn't require every switch hand ending with `break` statement.

File: src/savedialog.c [hunk 4]
```diff
@@ -175,6 +175,28 @@ gint game_compare_func(GtkTreeModel *model, GtkTreeIter *a, GtkTreeIter *b, gpoi
...
 
 
+static void save_load_dlg_response (GtkDialog * dlg, int response, gpointer user_data)
+{
+   switch (response)
+   {
+      case 100: /* save game */
+         do_save_game (GTK_WIDGET (user_data), dlg);
+         return;
+         break;
```
Comment (emorozov) [inline src/savedialog.c]: Same here with extra `break` statements.

File: src/savedialog.c [hunk 5]
```diff
@@ -175,6 +175,28 @@ gint game_compare_func(GtkTreeModel *model, GtkTreeIter *a, GtkTreeIter *b, gpoi
 }
 
 
+static void save_load_dlg_response (GtkDialog * dlg, int response, gpointer user_data)
+{
+   switch (response)
+   {
+      case 100: /* save game */
```
Comment (emorozov) [inline src/savedialog.c]: Does GTK define names for these constants?
They look undecipherable when used as anonymous magic constants.

**Review by emorozov — CHANGES_REQUESTED** [review]

File: src/savedialog.c (hunk 5 above)
Comment (wdlkmpx) [reply src/savedialog.c]: No, it's something you specify in `gtk_dialog_add_button(dialog, label, response)`

The "response" id is used in the dialog response callback.

response can be any GTK_RESPONSE_xxx but there is no GTK_RESPONSE_SAVE, so I specify an arbitrary number.

Comment (wdlkmpx) [reply src/savedialog.c]: It's also possible to #define stuff like GTK_RESPONSE_SAVE, GTK_RESPONSE_KILLBILL, etc, I didnt think about that at the moment. There are too few GTK_RESPONSE constants.

**Review by emorozov — APPROVED** [review]

**Comment by wdlkmpx** [comment]
It's probably mergeable now, the "force-push" diff provided by github
https://github.com/emorozov/gtkballs/compare/300badc12256e05e0155d4d988b3dfc2ad9b622f..3fb6dc4a9fa9191719d6475589da734fe21a120f

**Pull request merged**